src/main/scala/strawman/collection/immutable/Set.scala

```diff
@@ -22,4 +23,15 @@ trait SetMonoTransforms[A, +Repr]
 
   def - (elem: A): Repr
 
+}
+
+trait SetPolyTransforms[A, +C[X] <: Set[X] with SetLike[X, C]]
+  extends collection.SetPolyTransforms[A, C] {
+
+  protected def coll: C[A]
+
+  def ++ (that: collection.Set[A]): C[A] =
+    if (that.isEmpty) coll
+    else that.foldLeft[C[A]](coll)(_ + _)
```

Comment:
So, I ran the benchmarks and here is the result:

![concat](https://cloud.githubusercontent.com/assets/332812/24409086/7855a192-13cf-11e7-9a99-4f472f52f68e.png)

The best performance is achieved by the blue line (Scala 2.12 current `HashSet`). The green line show the performance of the above version of `++`, in our strawman, and the red line shows the performance of the `++` version inherited from [`IterablePolyTransforms`](https://github.com/scala/collection-strawman/blob/master/src/main/scala/strawman/collection/Iterable.scala#L189), based on `View.Concat` and `fromIterable`. Unfortunately, the version inherited from `Iterable` is ~2 times slower the one defined above…